### PR TITLE
Refactor subjects

### PR DESCRIPTION
**Pull request by classicalliu**
* Remove `AddressesUsedSubject` and make this progress sync
* Split `sync/utils` to `utils/array` and `utils/common` and update the type info
* Move `subscribe` from NodeService / WalletService / Subjects to listeners
* Move `checkAndSave` from `GetBlocks` to `SyncQueue`
* Update `getSubject()` method of subjects, check renderer here, and remove `isRenderer` check everywhere, delete `params / renderer-params` files.
* Move `genesisBlockHash / getChain` methods to `GetBlocks`
* Add `tipBlockNumber` to `NodeService` and remove `tipBlockListener` from `BlockListener` and `IndexerQueue`

### CODE REVIEW

File: packages/neuron-wallet/src/main.ts [hunk 1]
```diff
@@ -15,6 +16,7 @@ app.on('ready', async () => {
 
   NetworksService.getInstance().notifyAll()
   WalletService.getInstance().generateAddressesIfNecessary()
+  register()
```
Comment (Keith-CY) [inline packages/neuron-wallet/src/main.ts]: `registerListeners` could be more clear.

File: packages/neuron-wallet/src/models/subjects/address-created-subject.ts [hunk 2]
```diff
@@ -1,14 +1,16 @@
 import { ReplaySubject } from 'rxjs'
 import { Address } from 'database/address/address-dao'
+import { remote } from 'electron'
+import ProcessUtils from 'utils/process'
 
 export default class AddressCreatedSubject {
```
Comment (Keith-CY) [inline packages/neuron-wallet/src/models/subjects/address-created-subject.ts]: If is it possible to combine this subject into the address DB changed subject.

File: packages/neuron-wallet/src/main.ts (hunk 1 above)
Comment (classicalliu) [reply packages/neuron-wallet/src/main.ts]: updated

File: packages/neuron-wallet/src/utils/common.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+export default class CommonUtils {
+  public static sleep = (ms: number): Promise<void> => {
+    return new Promise(resolve => setTimeout(resolve, ms))
+  }
+
+  public static async retry<T>(times: number, interval: number, callback: () => T): Promise<T> {
+    let retryTime = 0
+
+    while (++retryTime < times) {
```
[the remark this answers is not in the document]
Comment (classicalliu) [reply packages/neuron-wallet/src/utils/common.ts]: Why ?

File: packages/neuron-wallet/src/models/subjects/address-created-subject.ts (hunk 2 above)
Comment (classicalliu) [reply packages/neuron-wallet/src/models/subjects/address-created-subject.ts]: Seems no need, different info / sender / listener

File: packages/neuron-wallet/src/utils/common.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Keith-CY) [reply packages/neuron-wallet/src/utils/common.ts]: > Why ?

Say `times = 3`

When `++retryTime`, it iterates `1, 2`
While `retryTime++` iterates `1, 2, 3`

[the remark this answers is not in the document]
Comment (classicalliu) [reply packages/neuron-wallet/src/utils/common.ts]: See last line in method, it's the 3rd time.

[the remark this answers is not in the document]
Comment (Keith-CY) [reply packages/neuron-wallet/src/utils/common.ts]: > See last line in method, it's the 3rd time.

Got, it's designed to throw an error if it fails after retrying.

**Comment by Keith-CY** [comment]
Please hold on until @ashchan approves

**Pull request merged**